File: 290-word-pattern.cpp
//https://leetcode.com/problems/word-pattern
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        istringstream ss(s);
        string word;
        unordered_map<string, char> m;
        unordered_map<char, string> m_rev;
        int index = 0;
        while(ss >> word){
            if(index == pattern.size())
                return false;
            
            if(m.find(word) != m.end()){
                if(m[word] != pattern[index])
                    return false;
            }
            if(m_rev.find(pattern[index]) != m_rev.end()){
                if(word != m_rev[pattern[index]])
                    return false;
            }
            m[word] = pattern[index];
            m_rev[pattern[index]] = word;

            index++;
        }
        return index == pattern.size();
    }
};

/*
"jquery"
"jquery"
"dog cat cat dog"
"a"
*/

File: 290-word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        l = s.split()
        if(len(l) != len(pattern)):
            return False
        m_ch_str = {}
        m_str_ch = {}
        for i in range(len(pattern)):
            if pattern[i] in m_ch_str and m_ch_str[pattern[i]] != l[i]:
                return False
            if l[i] in m_str_ch and m_str_ch[l[i]] != pattern[i]:
                return False
            else:
                m_ch_str[pattern[i]] = l[i]
                m_str_ch[l[i]] = pattern[i]
        
        return True

"""
"abba"
"dog dog dog dog"
"jquery"
"jquery"
"a"
"dog cat"
"""
